extension/src/time_series/pipeline/aggregation.rs

```diff
@@ -410,6 +412,213 @@ extension_sql!(r#"
+            PipelineThenCounterAgg {
+                num_elements: pipeline.0.num_elements,
+                elements: pipeline.0.elements,
+            }
+        }}
+    }
+
+    let mut elements = replace(pipeline.elements.as_owned(), vec![]);
+    elements.extend(then_counter_agg.elements.iter());
+    build! {
+        PipelineThenCounterAgg {
+            num_elements: elements.len().try_into().unwrap(),
+            elements: elements.into(),
+        }
+    }
+}
+
+#[pg_extern(
+    immutable,
+    parallel_safe,
+    name="counter_agg",
+    schema="toolkit_experimental"
+)]
+pub fn pipeline_counter_agg<'e>() -> toolkit_experimental::PipelineThenCounterAgg<'e> {
+    build! {
+        PipelineThenCounterAgg {
+            num_elements: 0,
+            elements: vec![].into(),
+        }
+    }
+}
+
+#[pg_extern(
+    immutable,
+    parallel_safe,
+    schema="toolkit_experimental"
+)]
+pub unsafe fn pipeline_counter_agg_support(input: Internal)
+-> Internal {
+    pipeline_support_helper(input, |old_pipeline, new_element| unsafe {
+        let new_element = PipelineThenCounterAgg::from_datum(new_element, false, 0)
+            .unwrap();
+        finalize_with_counter_agg(old_pipeline, new_element).into_datum().unwrap()
+    })
+}
+
+// using this instead of pg_operator since the latter doesn't support schemas yet
+// FIXME there is no CREATE OR REPLACE OPERATOR need to update post-install.rs
+//       need to ensure this works with out unstable warning
+extension_sql!(r#"
+ALTER FUNCTION toolkit_experimental."run_pipeline_then_counter_agg" SUPPORT toolkit_experimental.pipeline_counter_agg_support;
+
+CREATE OPERATOR -> (
+    PROCEDURE=toolkit_experimental."run_pipeline_then_counter_agg",
+    LEFTARG=toolkit_experimental.TimeSeries,
+    RIGHTARG=toolkit_experimental.PipelineThenCounterAgg
+);
+
+CREATE OPERATOR -> (
+    PROCEDURE=toolkit_experimental."finalize_with_counter_agg",
+    LEFTARG=toolkit_experimental.UnstableTimeseriesPipeline,
+    RIGHTARG=toolkit_experimental.PipelineThenCounterAgg
+);
+"#);
+
+pg_type! {
+    #[derive(Debug)]
+    struct PipelineThenHyperLogLog<'input> {
+        hll_size: u64,
+        num_elements: u64,
+        elements: [Element; self.num_elements],
+    }
+}
+
+ron_inout_funcs!(PipelineThenHyperLogLog);
+
+#[pg_extern(immutable, parallel_safe, schema="toolkit_experimental")]
+pub fn run_pipeline_then_hyperloglog<'s, 'p>(
+    mut timeseries: toolkit_experimental::TimeSeries<'s>,
+    pipeline: toolkit_experimental::PipelineThenHyperLogLog<'p>,
+) -> HyperLogLog<'static> {
+    unsafe {
+        timeseries = run_pipeline_elements(timeseries, pipeline.elements.iter());
+        let size: usize = pipeline.hll_size.try_into().unwrap();
+        let b = size.checked_next_power_of_two().unwrap().trailing_zeros();
+        let hasher = DatumHashBuilder::from_type_id(PgBuiltInOids::FLOAT8OID as u32, None);
+        let mut logger: HLL<pg_sys::Datum, DatumHashBuilder> = HLL::new(b as u8, hasher);
```

Comment:
I think this should be a function in hll; I think `DatumHashBuilder` is an internal implementation detail that shouldn't really be exposed...

Reply:
Hmm, I was thinking that `DatumHashBuilder` is eventually going to become its own public bit of infrastructure used in several different places (notably, I'm planning on making use of it for TopN).  However, I agree that here I'm using far too much HyperLogLog internal stuff, so I wrapped 533-542 into a `HyperLogLog::from_timeseries`.  I couldn't make it a proper `From` due to the size parameter, if you can think of a workaround there I'd love to see it.

Reply:
I kinda meant just `533-536` just a `Hyperloglog::with_size(size: usize, type: Oid)`; I don't think hll should need to know anything about timevectors much like timevectors shouldn't need to know anything about hll sizing contraints.